lib/model/channel.dart

```diff
@@ -192,12 +192,67 @@ class ChannelStoreImpl with ChannelStore {
 
       case ChannelDeleteEvent():
         for (final stream in event.streams) {
-          assert(identical(streams[stream.streamId], streamsByName[stream.name]));
-          assert(subscriptions[stream.streamId] == null
-            || identical(subscriptions[stream.streamId], streams[stream.streamId]));
-          streams.remove(stream.streamId);
-          streamsByName.remove(stream.name);
-          subscriptions.remove(stream.streamId);
+          final localStream = streams[stream.streamId];
+          if (localStream == null) continue; // TODO(log)
+
+          final ZulipStream(:streamId, name: String streamName) = localStream;
+          if (streamName != stream.name) {
+            // TODO(log)
+          }
+          assert(identical(streams[streamId], streamsByName[streamName]));
+          assert(subscriptions[streamId] == null
+            || identical(subscriptions[streamId], streams[streamId]));
+          streams.remove(streamId);
+          streamsByName.remove(streamName);
+          subscriptions.remove(streamId);
+        }
+
+      case ChannelUpdateEvent():
+        final stream = streams[event.streamId];
+        if (stream == null) return; // TODO(log)
+
+        final ZulipStream(:streamId, name: String streamName) = stream;
+        if (streamName != event.name) {
+          // TODO(log)
+        }
+        assert(identical(streams[streamId], streamsByName[streamName]));
```

Comment:
> This also doesn't do anything with `streamsByName` except in the `ChannelPropertyName.name` case. So this assert can be moved into that case.

Moved it to its specific case. But this assert in here actually helped me spot a bug in one of my previous revisions of this PR (actually when working on #886 which is stacked on top of this PR). In that revision, I forgot to include:
```diff
case ChannelPropertyName.name:
  stream.name = event.value as String;
+  streamsByName.remove(streamName);
+  streamsByName[stream.name] = stream;
```

So when first the channel `name` is changed and then some other property, i.e., `channelPostPolicy`, the assert would fail, simply because `streamsByName` was not updated accordingly. And now if we somehow have a bug where we don't update `streamsByName`, then the assert in the new location will not tell us about it. But now that it is covered in the tests, I think we're good to go. 🙂

Reply:
Yeah, that makes sense. I do think checking for that sort of inconsistency is useful. Two places we could do that that I think would work well (and which could coexist):
* We could use the same pattern as the `checkInvariants` function in `test/model/message_list_test.dart`: we'd write a function that scrubs the entire data structure and checks every aspect for consistency, and then have all the test cases in `test/model/channel_test.dart` arrange to call that function after every step they take. This is very thoroughly effective, for every situation that's exercised in those tests (even if the individual test doesn't think to check that aspect).
* We could arrange so that in the live app (in debug mode), every access to a channel or subscription in ChannelStore asserts that `_streams` and `_streamsByName` and `_subscriptions` are consistent with each other about that particular channel. In that case the asserts should get factored into a single method (with "debug" at the start of its name), which we can call from various places in `lib/model/channel.dart`. For accesses made from the rest of the app, we can extend [MapView](https://api.dart.dev/stable/3.5.1/dart-collection/MapView-class.html) (perhaps better yet, [UnmodifiableMapView](https://api.dart.dev/stable/3.5.1/dart-collection/UnmodifiableMapView-class.html)?) and add the asserts to `operator[]`.

But I'm also content to leave those aside for now.